crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -288,290 +353,268 @@ impl AppHandlerExecute for Ics20Transfer {
+) -> Result<()> {
+    if is_refund_source_zone(&asset, source_port, source_channel) {
         state
-            .increase_balance(recipient, &denom_trace, packet_amount)
+            .decrease_ibc_channel_balance(source_channel, &asset, amount)
+            .await
+            .context("failed to withdraw refund amount from ibc20 channel")?;
+        state
+            .increase_balance(recipient, &asset, amount)
             .await
             .context("failed to update user account balance in execute_ics20_transfer")?;
     } else {
         // register denomination in global ID -> denom map if it's not already there
+        // XXX(janis): is this correct? Shouldn't an asset that is being refunded always
+        // be present in the state if one attempted to transfer it out?
+        //
+        // TODO(janis): check if writing the asset to state if already present actually changes it.
+        // We could just always write it to state to avoid this extra check.
         if !state
-            .has_ibc_asset(&*trace_with_dest)
+            .has_ibc_asset(&asset)
             .await
             .context("failed to check if ibc asset exists in state")?
         {
             state
-                .put_ibc_asset(&trace_with_dest)
+                .put_ibc_asset(&asset)
                 .context("failed to put IBC asset in storage")?;
         }
 
         state
-            .increase_balance(recipient, &*trace_with_dest, packet_amount)
+            .increase_balance(recipient, asset, amount)
             .await
             .context("failed to update user account balance in execute_ics20_transfer")?;
     }
 
     Ok(())
 }
 
-/// Execute a refund for a failed ics20 transfer from a rollup to a remote IBC chain.
-///
-/// A withdrawal of tokens from a rollup to a remote IBC chain is started with a
-/// `Ics20Withdrawal` action via a (rollup's) bridge account on the sequencer.
-///
-/// This function then sends the tokens back to the rollup via a `Deposit` event,
-/// and again locks the tokens in the specified bridge account.
-///
-/// This function must only be called if the following conditions hold:
-/// 1. The ics20 tranfer is a refund;
-/// 2. The memo contained in the ics20 transfer packet can be parsed as a Ics20WithdrawalFromRollup.
-///
-/// The function then assumes that the `packet_data.sender` is the bridge account, and
-/// attempts to parse it as either a base-prefixed bech32m address, or as an ibc-compat-prefixed
-/// bech32 address. If the sender is an ibc-compat address, then it will be converted to its
-/// base-prefixed version.
-///
-/// The emitted deposit as seen by the rollup will *never* be the ibc-compat prefixed version.
 // TODO: Add `err` with https://github.com/astriaorg/astria/issues/1386 being done
 #[instrument(skip_all)]
-async fn execute_withdrawal_refund_to_rollup<S: StateWrite>(
-    state: &mut S,
-    packet_data: FungibleTokenPacketData,
+async fn refund_tokens_to_rollup<S: StateWrite>(
```

Comment:
also check `is_refund_source_zone` in this function?

Reply:
Can you double check? I was trying to be faithful to the original implementation, which did not do that check for these rollup-refunds.

Here (on current main) you can see that the refund to the rollup is short-circuiting on the memo parse:
https://github.com/astriaorg/astria/blob/1be156e4818f3234fb0a3f03257f33ca2627d95f/crates/astria-sequencer/src/ibc/ics20_transfer.rs#L423-L432

And here the final denom that will be put in the deposit is constructed:
https://github.com/astriaorg/astria/blob/1be156e4818f3234fb0a3f03257f33ca2627d95f/crates/astria-sequencer/src/ibc/ics20_transfer.rs#L570-L580

Reply:
I think this is correct now - refunding a source asset should result in unescrowing funds from the channel account as well, not just re-minting.